clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2163,6 +2163,51 @@ static void BuildFlattenedTypeList(QualType BaseTy,
   }
 }
 
+bool SemaHLSL::IsLineVectorLayoutCompatibleType(clang::QualType QT) {
+  if (QT.isNull())
+    return false;
+
+  llvm::SmallVector<QualType, 16> QTTypes;
+  BuildFlattenedTypeList(QT, QTTypes);
```

Comment:
This certainly works, but I wonder if building up the full flattened type is really needed. You just need to scan the type and make sure all elements are the same and get the total size while doing that. That should not require a separate list and allocations.

Reply:
Understandable, though I think this way is significantly simpler, and won't be much of a perf hit.
We allocate at most 4 elements in the list per resource, never any more for valid shaders. We would need to allocate at least one subelement with which to compare other subelements against using your method. It may be slightly less performant, but I think this is readable and it doesn't drag the performance down from O(1) to O(N) for valid shaders.(N being number of subelements)

Reply:
Well, QTTypes is initialized with size 16, and as you say it should be 4 here.

Note that BuildFlattenedTypeList is creating 3 additional SmallVectors instances for WorkList, FieldTypes and ElementFields (though this last one is an invalid case for this builtin), and doing additional work making sure the types are in the right order.

It is certainly simpler to let BuildFlattenedTypeList do its work and then evaluate the type trait based on the results. And in this case, it might be fine because the trait will most likely be evaluated just once for each template instantiation of a buffer class & element type pair.

It is however important to know what is behind helper functions like `BuildFlattenedTypeList` to make sure it is not an overkill for a given task.